### ISSUE TITLE
What is the best cloud-based automated UI test (Selenium) runner?

### ISSUE THREAD

**Issue by theresaanna** [khandelwal/AskShashank#25]
Dearest Shashank, 

You are welcome to answer this question, but I know that you claim to stay as far away from JavaScripty things as possible (even though we all know that your weekend side projects are in Coffeescript). Really, I'm using this as a vehicle to ask our estranged brethren @ascott1 and @contolini.

I like Sauce Labs, but I started to set it up and decided to hold off instead. Is there anything better out there? TBH, I don't really want to use Grunt. Having tasks split between NPM and Grunt lands is unsavory to me and, of the two, I find NPM's handling far more graceful. Things I don't like about Sauce: having to do extra work to get pass/fail feedback on your test running client, that they always seem to be scrambling to keep docs and tooling up to date, that often times the docs aren't checked for validity (twice I've found this with setup docs). I guess it doesn't even have to be cloud-based, as long as its stupidly simple and reliable to set up on multiple environments in an automated fashion (think Travis). That seems impossible with Selenium in the equation, though.

I don't mind dealing with Sauce's imperfections if its the best of what we've got. But, @ascott1 and @contolini, y'all are way cooler with me with your twinsie JS powers and usually know of things before I do.

Please, Shashank, have mercy on my poor, JS and NPMing soul and keep this issue active. @rjmajma, I'm asking for OpenFEC, so can you bribe Shashank if necessary? Thanks.

<3,
T

**Comment by ascott1** [comment]
Shashank, this thread is below your enlightened state. Please don't feel obligated to read on.

---

In Shashank's world there are no interfaces, there is only text. Much like Neo staring into the Matrix, Vim (set to a lovely shade of orange, of course) with a web browser plugin provides Shashank with all of the interface he ever needs. 

Unfortunately, for us humans interfaces are required for us to interact with machines in a meaningful way. While our dear friend Shashank may struggle to see their value, some of us lower class developers are tasked with building the portion of applications that us commoners must interact with and we would even like to ensure that these inferior bits of the tool work as expected.

All of that aside, I have yet to find a decent alternative to Sauce. [Browserling](https://browserling.com/) looks promising, but certainly isn't as fully featured.

**Comment by contolini** [comment]
Last I checked, Sauce requires you to create a separate user account for every open source repo you're testing. This was a deal breaker for me because at the time we were projectile vomiting node modules and creating new repos constantly.

I've been meaning to look into BrowserStack. Their [JS API is free for OOS](http://www.browserstack.com/javascript-testing-api). If you try it, please report back.

**Comment by theresaanna** [comment]
@ascott1 @contolini I super love you guys. Thanks. I'll take a look at BrowserStack and report back on what I decide and why. 

So I'm not publishing modules I write cause they're project-specific, so the Sauce account thing isn't an issue.

**Comment by theresaanna** [comment]
BrowserStack dealbreaker: Selenium tests aren't free.
Even if I were going to pay, Sauce has screen recordings of tests running, which I've found to be very valuable in debugging Selenium test runs.
Pro-BrowserStack: the tooling around it looks quite nice.

**Comment by jpyuda** [comment]
From the makers of Browserling: https://ci.testling.com

**Comment by contolini** [comment]
Testing is hands-down the best as you don't even need an account for open source repos buuuuut it's been down for months. https://twitter.com/testling

**Comment by ascott1** [comment]
Yeah, @contolini's response is why I didn't recommend Testling, which _is_ awesome (when it works....)

**Comment by sfarthin** [comment]
I was taking look at CFPB's Github activity and came across thread which I think I have a lot to contribute to. I am hoping to join the CFPB front-end developer team in 2015.

You certainly do not need a task runner (Grunt, Gulp, etc) to run javascript tests in browsers with karma. I [cooked up a number of examples in this repository](https://github.com/sfarthin/js-and-selenium) which only require a simple configuration file (karma.conf.js) and can be configured to run with "npm test". These examples run in a single browser on the local machine and do not need a selenium server running separately, but karma can also be configured to connect to a remote selenium server on another machine/VM or run on multiple browsers.

I am guessing [this example](https://github.com/sfarthin/js-and-selenium/tree/master/karma-jasmine-phantomjs-browserify) will be most applicable because it demonstrates how to incorporate browserify.

For screenshots and more direct access to selenium's methods check out the [Protractor](https://github.com/sfarthin/js-and-selenium/tree/master/protractor-mocha-chrome) and [Webdriverjs](https://github.com/sfarthin/js-and-selenium/tree/master/webdriverjs-tape-remoteIE) examples.

Hope this can be useful.
